Update intermediate tutorial, contributing.md; add template.md (#461)
smartcontractkit/documentation: TEMPLATE.md

---
layout: nodes.liquid
section: smartContract
date: Last Modified
title: "[insert title here]"
permalink: "[insert path here]"
excerpt: "Smart Contracts and Chainlink"
whatsnext: {"[insert hyperlink title]":"insert path to hyperlink", "":""}
metadata:
  image:
    0: "insert image here"
---

> *This doc is meant to be used as a template for creating a tutorial doc. Users should emulate the structure of this document when creating their own templates. The content of this doc comes from beginners-tutorial.md.*

Reviewer: ```suggestion
*This doc is meant to be used as a template for creating a tutorial doc. Contributors should emulate the structure of this document when creating new pages. The content of this doc comes from beginners-tutorial.md.*
```

> *Prerequisites should have its own block at the beginning of the tutorial. Make sure to tell the user what knowledge they should already have and where to go if they need previous reading. Make sure to direct them to other docs pages.*

Reviewer: ```suggestion
*Prerequisites should have its own block at the beginning of the tutorial. Tell the user what tools are required, what knowledge they should already have, and where to go if they need previous reading. Provide links to the other docs pages.*
```

> > 👍 Prerequisites

Reviewer: Let's standardize on "Requirements" instead of "Prerequisites" unless you have a preference here.

>
> This tutorial requires .... If you're unfamiliar with these concepts, follow ...
> or 
> New to []? We'll walk you through...

*If a tutorial has a corresponding YouTube tutorial, link it at the beginning of the page.*
<p>
  https://www.youtube.com/watch?v=rFXSEEQG9YE
</p>

*If a portion of the YouTube video is deprecated or out-of-date, add a note alerting the user.*

> 🚧 Note
>

> > The video uses a seed phrase to request randomness. This feature has been deprecated. Please refer to the code in this tutorial for the most up to date procedures.

Reviewer: ```suggestion
> The video uses a seed phrase to request randomness. This feature has been deprecated. Refer to the code in this tutorial for the most up to date procedures.
```

> # Overview <!-- omit in toc -->

Reviewer: ```suggestion
# Overview
```

> *Write 1-4 sentences on what this tutorial covers. Focus on the end result of the tutorial and the overarching concepts a user will learn about/implement. Throughout the entire piece, make sure to use second person singular (you). This should be consistent throughout the entire document, but exceptions can be made. Avoid passive voice as well. Remember, anyone reading the tutorial should interpret it as a source of truth. Make sure to write in a voice which is factual and objective.*

Reviewer: ```suggestion
*Write 1-4 sentences on what this tutorial covers. Focus on the end result of the tutorial and the overarching concepts a user will learn and implement. Throughout the entire document, use second person singular (you). This should be consistent throughout the entire document, but exceptions can be made. Avoid passive voice as well. Remember, anyone reading the tutorial should interpret it as a source of truth. Make sure to write in a voice which is factual and objective.*
```

A valuable feature of smart contracts is that they can store and manage on-chain assets (like [ETH or ERC20 tokens](https://ethereum.org/en/developers/docs/standards/tokens/erc-20/)), just like you can with an Ethereum wallet. Because they have an on-chain address like a wallet, they can do everything any other address can. This opens the door for programming automated actions when receiving and transferring assets.

*Notice how users are given external links to read more about concepts such as ETH or ERC20 tokens. They do not have to undertake a separate search.*

Smart contracts can connect to real-world market prices of assets to produce powerful applications. Chainlink's **[Data Feeds](../using-chainlink-reference-contracts/)** feature allows users to quickly and securely connect smart contracts to such assets in a single call. 

# 2. What language is a smart contract written in?

...

# 3. What does a smart contract look like?

> *For code samples, be thorough. Redundancy is better than assuming the user have sufficient knowledge on a certain topic. Create markdown code samples wherever  applicable. Make sure to give context as to what each code sample is doing.*

Reviewer: ```suggestion
*For code samples, be thorough. Redundancy is better than assuming the user have sufficient knowledge on a certain topic. Create markdown code samples wherever  applicable. Provide a detailed explanation about what each code sample does.*
```

```solidity
pragma solidity 0.8.7;

contract HelloWorld {
    string public message;

    constructor(string memory initialMessage) {
        message = initialMessage;
    }

    function updateMessage(string memory newMessage) public {
        message = newMessage;
    }
}
```

*Subheadings don't need to be numbered. If a step of the tutorial doesn't need its own heading, it's best to make a subheading. Example:*

## Define the Version

The first thing that every Solidity file must have is the Solidity version definition. The version HelloWorld.sol is using is 0.8.7, defined by `pragma solidity 0.8.7;`

You can see the latest versions of the Solidity compiler [here](https://github.com/ethereum/solc-bin/blob/gh-pages/bin/list.txt/?target=_blank). You may also notice Solidity files containing definitions with multiple versions of Solidity:

```solidity
pragma solidity >=0.7.0 <0.9.0;
```
This means that the code is written for Solidity version 0.7.0, or a newer version of the language up to, but not including version 0.9.0. In short, `pragma` is used to instruct the compiler as how to treat the code.

...

> *As a final step, make sure to include further reading with relevant blog posts. Additionally, make sure to link users to other-tutorials.md. Example:*

Reviewer: ```suggestion
*As a final step, include further reading with relevant blog posts and link users to other tutorials. Example:*
```

# 4. Further Reading

> To read more about data feeds, read our blog posts:

Reviewer: ```suggestion
To learn more about data feeds, read the Data Feed blog posts:
```

- [Build a dApp on xDai Chain With Secure Data Feeds](https://blog.chain.link/build-a-dapp-on-xdai-chain-with-secure-data-feeds/)
- [Build a dApp on Binance Smart Chain With Secure Data Feeds](https://blog.chain.link/build-a-dapp-on-binance-smart-chain-with-secure-data-feeds/)
- [The 3 Levels of Data Aggregation in Chainlink Price Feeds](https://blog.chain.link/levels-of-data-aggregation-in-chainlink-price-feeds/)

> To explore more applications of data feeds, check out our [other tutorials](/docs/other-tutorials/#data-feeds-applications).

Reviewer: ```suggestion
To explore more applications for data feeds, check out our [other tutorials](/docs/other-tutorials/#data-feeds-applications).
```
